### PR TITLE
KAFKA-15853: Move TransactionLogConfig and TransactionStateManagerConfig getters out of KafkaConfig

### PR DESCRIPTION
**Pull request by OmniaGM**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mimaison** [comment]
It looks like this breaks a bunch of tests:
```
java.lang.NullPointerException
	at kafka.server.KafkaConfig.get(KafkaConfig.scala:212)
	at org.apache.kafka.common.config.AbstractConfig.getInt(AbstractConfig.java:177)
```

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java [hunk 1]
```diff
@@ -61,15 +62,67 @@ public final class TransactionLogConfigs {
...
+            .define(TRANSACTION_PARTITION_VERIFICATION_ENABLE_CONFIG, BOOLEAN, TRANSACTION_PARTITION_VERIFICATION_ENABLE_DEFAULT, LOW, TRANSACTION_PARTITION_VERIFICATION_ENABLE_DOC)
 
-            .define(TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_CONFIG, INT, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_DEFAULT, atLeast(1), LOW, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_MS_DOC)
+            .define(PRODUCER_ID_EXPIRATION_MS_CONFIG, INT, PRODUCER_ID_EXPIRATION_MS_DEFAULT, atLeast(1), LOW, PRODUCER_ID_EXPIRATION_MS_DOC)
             // Configuration for testing only as default value should be sufficient for typical usage
-            .defineInternal(TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_CONFIG, INT, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DEFAULT, atLeast(1), LOW, TransactionLogConfigs.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DOC);
+            .defineInternal(PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_CONFIG, INT, PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DEFAULT, atLeast(1), LOW, PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DOC);
+
+    private AbstractConfig config;
```
Comment (chia7712) [inline transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: please add `final`

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java [hunk 2]
```diff
@@ -61,15 +62,67 @@ public final class TransactionLogConfigs {
...
+    public int transactionTopicPartitions() {
+        return transactionTopicPartitions;
+    }
+
+    public int transactionTopicSegmentBytes() {
+        return transactionTopicSegmentBytes;
+    }
+
+    public Boolean transactionPartitionVerificationEnable() {
+        return config.getBoolean(TRANSACTION_PARTITION_VERIFICATION_ENABLE_CONFIG);
```
Comment (chia7712) [inline transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:121]: could you please add comments to for those dynamic configs?

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionStateManagerConfig.java [hunk 3]
```diff
@@ -49,9 +50,38 @@ public final class TransactionStateManagerConfigs {
...
-            .define(TransactionStateManagerConfigs.TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONAL_ID_CLEANUP_INTERVAL_MS_CONFIG, INT, TransactionStateManagerConfigs.TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONAL_ID_CLEANUP_INTERVAL_MS_DEFAULT, atLeast(1), LOW, TransactionStateManagerConfigs.TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONS_INTERVAL_MS_DOC);
+            .define(TRANSACTIONAL_ID_EXPIRATION_MS_CONFIG, INT, TRANSACTIONAL_ID_EXPIRATION_MS_DEFAULT, atLeast(1), HIGH, TRANSACTIONAL_ID_EXPIRATION_MS_DOC)
+            .define(TRANSACTIONS_MAX_TIMEOUT_MS_CONFIG, INT, TRANSACTIONS_MAX_TIMEOUT_MS_DEFAULT, atLeast(1), HIGH, TRANSACTIONS_MAX_TIMEOUT_MS_DOC)
+            .define(TRANSACTIONS_ABORT_TIMED_OUT_TRANSACTION_CLEANUP_INTERVAL_MS_CONFIG, INT, TRANSACTIONS_ABORT_TIMED_OUT_TRANSACTION_CLEANUP_INTERVAL_MS_DEFAULT, atLeast(1), LOW, TRANSACTIONS_ABORT_TIMED_OUT_TRANSACTIONS_INTERVAL_MS_DOC)
+            .define(TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONAL_ID_CLEANUP_INTERVAL_MS_CONFIG, INT, TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONAL_ID_CLEANUP_INTERVAL_MS_DEFAULT, atLeast(1), LOW, TRANSACTIONS_REMOVE_EXPIRED_TRANSACTIONS_INTERVAL_MS_DOC);
+
+    private int transactionalIdExpirationMs;
```
Comment (chia7712) [inline transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionStateManagerConfig.java]: please add `final`

**Review by chia7712 — COMMENTED** [review]
@OmniaGM thanks for this nice refactor. some trivial comments are left. Also, please take a look at failed tests.

**Comment by chia7712** [comment]
@OmniaGM I love this PR, but it includes too many files. Hence, I'd like to delay this PR until 3.9.0 gets release :)

**Comment by mimaison** [comment]
So is this the format we're going for? I see we have a constructor accepting `AbstractConfig` and getters for all fields, plus potentially a `validate()` method for any validations that only rely on configs in the class.

I think it would be good to document the expected target and explain the pros and cons vs alternatives in the Jira.

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java (hunk 2 above)
Comment (mimaison) [reply transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:121]: Even with the comments, I'm not sure I understand why this is not initialized in the constructor.

Can you explain how this is supposed to work? AbstractConfig instances are kind of read only by default but it looks like we're expecting it to change here. Am I missing something?

Comment (chia7712) [reply transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:121]: we had a discussion for that before (https://github.com/apache/kafka/pull/16458#discussion_r1662130270)

`AbstractConfig` is read-only, but the sub class `KafkaConfig` can update inner configs dynamically (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaConfig.scala#L192). Hence, the passed `AbstractConfig` always have "up-to-date configs"

There is a jira which tries to refactor it (https://issues.apache.org/jira/browse/KAFKA-17001), but I haven't worked on it because that will introduce major changes to scala code.

In short, the basic rules for now are shown below.

1. the non-dynamic config will be evaluated in constructor.
2. the dynamic config will be  evaluated in getter (as `AbstractConfig` always have latest configs)

Comment (mimaison) [reply transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:121]: Thanks for the explanations!

**Comment by chia7712** [comment]
@OmniaGM could you please fix the conflicts? thanks!

**Comment by chia7712** [comment]
@OmniaGM Both `TransactionStateManagerConfig` and `TransactionLogConfig` are small file, and they are normally called together. Hence, do you consider merging them into one file?

**Comment by OmniaGM** [comment]
> @OmniaGM Both `TransactionStateManagerConfig` and `TransactionLogConfig` are small file, and they are normally called together. Hence, do you consider merging them into one file?

They are called together only `KafkaConfig`, `TransactionStateManager`, `TransactionCoordinator`, `LogManager`. While `TransactionLogConfig` is used in `DynamicBrokerConfig`, `KafkaServer`,  `AutoTopicCreationManager`, `ReplicaManager`, `BrokerMetadataPublisher`. I split them into 2 previously because the comment on them in KafkaConfig back then suggested they are different domains. I guess we can move `TransactionLogConfig` into `TransactionStateManagerConfig`. I just feel like the scope of where each one is used is bit different but generally the code access them from `KafkaConfig` anyway.

**Comment by chia7712** [comment]
>  I guess we can move TransactionLogConfig into TransactionStateManagerConfig. I just feel like the scope of where each one is used is bit different but generally the code access them from KafkaConfig anyway.

I have no strong reason to merge them though 😃 

Let's keep current style. I will take another look later

**Comment by chia7712** [comment]
@OmniaGM Sorry that could you please fix the conflicts again? This PR will have highest priority in my queue

**Review by chia7712 — APPROVED** [review]
